### PR TITLE
Convert ra-ui-materialui inputs, fields & layouts to typescript

### PR DESCRIPTION
**Pull request by josephktcheung**
1. Convert the following components to typescript:
    - `Link`
    - `ReferenceArrayField`
    - `ReferenceField`
    - `ReferenceManyField`
    - `ResettableTextField`
    - `AppBar`
    - `CardContentInner`
    - `DeviceTestWrapper`
    - `Error`
    - `Layout`
    - `LinearProgress`
    - `Loading`
    - `LoadingIndicator`
    - `NotFound`
    - `Notification`
    - `Responsive`
    - `Sidebar`
    - `Title`
    - `TitleForRecord`
    - `TopToolbar`
    - `UserMenu`
2. Export types from `ra-core`

### CODE REVIEW

File: packages/ra-ui-materialui/src/layout/AppBar.tsx [hunk 1]
```diff
@@ -61,7 +62,7 @@ const AppBar = ({
 }) => {
     const classes = useStyles({ classes: classesOverride });
     const dispatch = useDispatch();
-    const isXSmall = useMediaQuery(theme => theme.breakpoints.down('xs'));
+    const isXSmall = useMediaQuery<any>(theme => theme.breakpoints.down('xs'));
```
Comment (djhi) [inline packages/ra-ui-materialui/src/layout/AppBar.tsx]: Why are you using `any` here ?

```suggestion
    const isXSmall = useMediaQuery<Theme>(theme => theme.breakpoints.down('xs'));
```

Comment (djhi) [reply packages/ra-ui-materialui/src/layout/AppBar.tsx]: Don't forget to import the Theme type

**Comment by josephktcheung** [comment]
> Thanks! Would you mind making smaller pull requests next time? The smaller they are, the easier they are to review.

Sure, will do on my next PR

File: packages/ra-ui-materialui/src/field/ReferenceArrayField.tsx [hunk 2]
```diff
@@ -1,9 +1,28 @@
...
-import { useReferenceArrayFieldController } from 'ra-core';
+import {
+    useReferenceArrayFieldController,
+    Identifier,
+    RecordMap,
+} from 'ra-core';
 import { fieldPropTypes } from './types';
+import { ClassNameMap } from '@material-ui/styles';
+
+interface ReferenceArrayFieldProps {
```
Comment (fzaninotto) [inline packages/ra-ui-materialui/src/field/ReferenceArrayField.tsx]: I think ra-ui-materialui should publish a generic FieldProps interface, which each individual field should extend.

File: packages/ra-ui-materialui/src/field/ReferenceArrayField.tsx [hunk 3]
```diff
@@ -80,7 +103,22 @@ const useStyles = makeStyles(
     { name: 'RaReferenceArrayField' }
 );
 
-export const ReferenceArrayFieldView = ({
+type ReferenceArrayFieldClassKey = 'progress';
```
Comment (fzaninotto) [inline packages/ra-ui-materialui/src/field/ReferenceArrayField.tsx]: Where does this come from?

File: packages/ra-ui-materialui/src/field/ReferenceArrayField.tsx [hunk 4]
```diff
@@ -1,9 +1,28 @@
...
+} from 'ra-core';
 import { fieldPropTypes } from './types';
+import { ClassNameMap } from '@material-ui/styles';
+
+interface ReferenceArrayFieldProps {
+    addLabel?: boolean;
+    allowEmpty: boolean;
+    basePath: string;
+    children: ReactElement;
+    classes: Partial<ClassNameMap<ReferenceArrayFieldClassKey>>;
```
Comment (fzaninotto) [inline packages/ra-ui-materialui/src/field/ReferenceArrayField.tsx]: not required

File: packages/ra-ui-materialui/src/field/ReferenceArrayField.tsx [hunk 5]
```diff
@@ -1,9 +1,28 @@
...
+    useReferenceArrayFieldController,
+    Identifier,
+    RecordMap,
+} from 'ra-core';
 import { fieldPropTypes } from './types';
+import { ClassNameMap } from '@material-ui/styles';
+
+interface ReferenceArrayFieldProps {
+    addLabel?: boolean;
+    allowEmpty: boolean;
```
Comment (fzaninotto) [inline packages/ra-ui-materialui/src/field/ReferenceArrayField.tsx]: not required

File: packages/ra-ui-materialui/src/field/ReferenceArrayField.tsx [hunk 6]
```diff
@@ -1,9 +1,28 @@
...
+    basePath: string;
```
Comment (fzaninotto) [inline packages/ra-ui-materialui/src/field/ReferenceArrayField.tsx]: not required

File: packages/ra-ui-materialui/src/field/ReferenceArrayField.tsx [hunk 7]
```diff
@@ -1,9 +1,28 @@
...
+    children: ReactElement;
+    classes: Partial<ClassNameMap<ReferenceArrayFieldClassKey>>;
+    className: string;
```
Comment (fzaninotto) [inline packages/ra-ui-materialui/src/field/ReferenceArrayField.tsx]: not required

File: packages/ra-ui-materialui/src/field/ReferenceArrayField.tsx [hunk 8]
```diff
@@ -1,9 +1,28 @@
...
+    label: string;
```
Comment (fzaninotto) [inline packages/ra-ui-materialui/src/field/ReferenceArrayField.tsx]: not required

File: packages/ra-ui-materialui/src/field/ReferenceArrayField.tsx [hunk 9]
```diff
@@ -1,9 +1,28 @@
...
+    reference: string;
+    resource: string;
```
Comment (fzaninotto) [inline packages/ra-ui-materialui/src/field/ReferenceArrayField.tsx]: not required

File: packages/ra-ui-materialui/src/field/ReferenceArrayField.tsx [hunk 10]
```diff
@@ -80,7 +103,22 @@ const useStyles = makeStyles(
     { name: 'RaReferenceArrayField' }
 );
 
-export const ReferenceArrayFieldView = ({
+type ReferenceArrayFieldClassKey = 'progress';
+
+interface ReferenceArrayFieldViewProps {
+    children: ReactElement;
+    classes?: Partial<ClassNameMap<ReferenceArrayFieldClassKey>>;
+    className: string;
```
Comment (fzaninotto) [inline packages/ra-ui-materialui/src/field/ReferenceArrayField.tsx]: not required

File: packages/ra-ui-materialui/src/field/ReferenceArrayField.tsx [hunk 11]
```diff
@@ -80,7 +103,22 @@ const useStyles = makeStyles(
...
+    data: RecordMap;
```
Comment (fzaninotto) [inline packages/ra-ui-materialui/src/field/ReferenceArrayField.tsx]: not required (while loading, the data is empty)

File: packages/ra-ui-materialui/src/field/ReferenceArrayField.tsx [hunk 12]
```diff
@@ -80,7 +103,22 @@ const useStyles = makeStyles(
...
+    error?: any;
+    ids: Identifier[];
```
Comment (fzaninotto) [inline packages/ra-ui-materialui/src/field/ReferenceArrayField.tsx]: not required (while loading, it's empty)

File: packages/ra-ui-materialui/src/field/ReferenceArrayField.tsx [hunk 13]
```diff
@@ -107,9 +145,9 @@ export const ReferenceArrayFieldView = ({
 };
 
 ReferenceArrayFieldView.propTypes = {
-    classes: PropTypes.object,
+    classes: PropTypes.any,
     className: PropTypes.string,
-    data: PropTypes.object,
+    data: PropTypes.any,
```
Comment (fzaninotto) [inline packages/ra-ui-materialui/src/field/ReferenceArrayField.tsx]: why did you change this one?

File: packages/ra-ui-materialui/src/field/ReferenceArrayField.tsx (hunk 3 above)
Comment (josephktcheung) [reply packages/ra-ui-materialui/src/field/ReferenceArrayField.tsx]: This is from key of the [`makeStyle` object in line 78](https://github.com/marmelab/react-admin/pull/4448/files/36635530bed3b6d66f1888d1b19715b64a3a6bf7#diff-edf706d49e4f05c200bd721a16f0334bR101) and it's similar to what material-ui is doing with its components
e.g. [`AppBar`](https://github.com/mui-org/material-ui/blob/master/packages/material-ui/src/AppBar/AppBar.d.ts#L17-L26). It'd allow library user to know which classes this component they can override

File: packages/ra-ui-materialui/src/field/ReferenceArrayField.tsx (hunk 2 above)
Comment (josephktcheung) [reply packages/ra-ui-materialui/src/field/ReferenceArrayField.tsx]: Agree. I wonder if the list here has included all generic props.

File: packages/ra-ui-materialui/src/field/ReferenceArrayField.tsx (hunk 3 above)
Comment (josephktcheung) [reply packages/ra-ui-materialui/src/field/ReferenceArrayField.tsx]: [`Placeholder.tsx`](https://github.com/marmelab/react-admin/blob/master/packages/ra-ui-materialui/src/list/Placeholder.tsx#L17) is using `Record<'root', string>` for `classes` prop,
while `ClassNameMap` is string a type alias of `Record` that sets default type to `string` from material-ui
```typescript
ClassNameMap<ClassKey extends string = string> = Record<ClassKey, string>
```

Comment (josephktcheung) [reply packages/ra-ui-materialui/src/field/ReferenceArrayField.tsx]: I name it `ReferenceArrayFieldClassKey` so that I can reuse it in [line 110](https://github.com/marmelab/react-admin/pull/4448/files/36635530bed3b6d66f1888d1b19715b64a3a6bf7#diff-edf706d49e4f05c200bd721a16f0334bR110)

File: packages/ra-ui-materialui/src/field/ReferenceArrayField.tsx (hunk 13 above)
Comment (josephktcheung) [reply packages/ra-ui-materialui/src/field/ReferenceArrayField.tsx]: Because `RecordMap` from `ra-core` has conflict with `PropTypes.object` and I'm not sure how to resolve it yet.

Comment (josephktcheung) [reply packages/ra-ui-materialui/src/field/ReferenceArrayField.tsx]: Here's the error I got from ts:
```
Type 'Requireable<object>' is not assignable to type 'Validator<RecordMap<Record>>'.
  The types of '[nominalTypeHack].type' are incompatible between these types.
    Type 'object' is not assignable to type 'RecordMap<Record>'.
      Index signature is missing in type '{}'.
```

File: packages/ra-ui-materialui/src/field/ReferenceArrayField.tsx (hunk 2 above)
Comment (josephktcheung) [reply packages/ra-ui-materialui/src/field/ReferenceArrayField.tsx]: Found `FieldProps` in `types.ts`, will use it to extend `ReferenceArrayFieldProps`

File: packages/ra-ui-materialui/src/field/ReferenceArrayField.tsx (hunk 9 above)
Comment (josephktcheung) [reply packages/ra-ui-materialui/src/field/ReferenceArrayField.tsx]: `reference`, `resource` are required by `useReferenceArrayFieldController`

File: packages/ra-ui-materialui/src/field/ReferenceArrayField.tsx (hunk 6 above)
Comment (josephktcheung) [reply packages/ra-ui-materialui/src/field/ReferenceArrayField.tsx]: `basePath`is required by `useReferenceArrayFieldController`

File: packages/ra-ui-materialui/src/field/ReferenceArrayField.tsx (hunk 12 above)
Comment (josephktcheung) [reply packages/ra-ui-materialui/src/field/ReferenceArrayField.tsx]: Return props of [`useReferenceArrayFieldController`](https://github.com/marmelab/react-admin/blob/master/packages/ra-core/src/controller/field/useReferenceArrayFieldController.ts#L18-L19) which is injected into `ReferenceArrayFieldView` has the following interface:
```typescript
interface ReferenceArrayProps {
    ids: Identifier[];
    data: RecordMap;
    error?: any;
    loading: boolean;
    loaded: boolean;
    referenceBasePath: string;
}
```

File: packages/ra-ui-materialui/src/field/ReferenceArrayField.tsx (hunk 9 above)
Comment (josephktcheung) [reply packages/ra-ui-materialui/src/field/ReferenceArrayField.tsx]: The thing is these props may be injected by a parent, will think how to deal with it

File: packages/ra-ui-materialui/src/field/ReferenceArrayField.tsx (hunk 11 above)
Comment (josephktcheung) [reply packages/ra-ui-materialui/src/field/ReferenceArrayField.tsx]: `useRefereneArrayFieldController` returns `data: RecordMap;` in its `ReferenceArrayProps`

File: packages/ra-ui-materialui/src/field/ReferenceArrayField.tsx (hunk 4 above)
Comment (josephktcheung) [reply packages/ra-ui-materialui/src/field/ReferenceArrayField.tsx]: Fixed

**Comment by josephktcheung** [comment]
Dear all, I'm closing this PR in favour of smaller PRs e.g. #4459. I've made the changes @fzaninotto suggested in the review comments. Please continue our discussion in #4459, thanks.
